src/Microsoft.Diagnostics.Monitoring.WebApi/Utilities.cs

```diff
@@ -46,6 +55,78 @@ public static KeyValueLogScope CreateArtifactScope(string artifactType, IEndpoin
             return scope;
         }
 
+        public static AggregateSourceConfiguration GetTraceConfiguration(Models.TraceProfile profile, int metricsIntervalSeconds)
+        {
+            var configurations = new List<MonitoringSourceConfiguration>();
+            if (profile.HasFlag(Models.TraceProfile.Cpu))
+            {
+                configurations.Add(new CpuProfileConfiguration());
+            }
+            if (profile.HasFlag(Models.TraceProfile.Http))
+            {
+                configurations.Add(new HttpRequestSourceConfiguration());
+            }
+            if (profile.HasFlag(Models.TraceProfile.Logs))
+            {
+                configurations.Add(new LoggingSourceConfiguration(
+                    LogLevel.Trace,
+                    LogMessageType.FormattedMessage | LogMessageType.JsonMessage,
+                    filterSpecs: null,
+                    useAppFilters: true));
+            }
+            if (profile.HasFlag(Models.TraceProfile.Metrics))
+            {
+                configurations.Add(new MetricSourceConfiguration(metricsIntervalSeconds, Enumerable.Empty<string>()));
+            }
+
+            return new AggregateSourceConfiguration(configurations.ToArray());
+        }
+
+        public static EventPipeProviderSourceConfiguration GetCustomTraceConfiguration(Models.EventPipeProvider[] configurationProviders, bool requestRundown, int bufferSizeInMB)
+        {
+            var providers = new List<EventPipeProvider>();
+
+            foreach (Models.EventPipeProvider providerModel in configurationProviders)
+            {
+                if (!IntegerOrHexStringAttribute.TryParse(providerModel.Keywords, out long keywords, out string parseError))
+                {
+                    throw new InvalidOperationException(parseError);
+                }
+
+                providers.Add(new EventPipeProvider(
+                    providerModel.Name,
+                    providerModel.EventLevel,
+                    keywords,
+                    providerModel.Arguments
+                    ));
+            }
+
+            return new EventPipeProviderSourceConfiguration(
+                providers: providers.ToArray(),
+                requestRundown: requestRundown,
+                bufferSizeInMB: bufferSizeInMB);
+        }
+
+        public static async Task GetTraceAction(IEndpointInfo endpointInfo, MonitoringSourceConfiguration configuration, TimeSpan duration, Stream outputStream, CancellationToken token)
```

Comment:
One small thing here - this gets called from within `DiagController`, which doesn't have the `startCompletionSource`. My current implementation just has `DiagController` passing null for that parameter and then this method doing a null check before `startCompletionSource.TrySetResult(null)`, but I wasn't sure if you'd consider that messy (since `token` is the last parameter, we can't give `startCompletionSource` a default null value, so I'm explicitly passing `null` for now)

Reply:
That's absolutely fine to do. The HTTP requests don't care when it started, only that it finishes.